src/panels/media-browser/ha-panel-media-browser.ts

```diff
@@ -0,0 +1,160 @@
+import { showSelectMediaSourceDialog } from "./show-select-media-source-dialog";
+
+@customElement("ha-panel-media-browser")
+class PanelMediaBrowser extends LitElement {
+  @property({ attribute: false }) public hass!: HomeAssistant;
+
+  @property({ type: Boolean, reflect: true })
+  public narrow!: boolean;
+
+  // @ts-ignore
+  @LocalStorage("mediaBrowseEntityId")
+  private _entityId = BROWSER_SOURCE;
+
+  protected render(): TemplateResult {
+    const stateObj = this._entityId
+      ? this.hass.states[this._entityId]
+      : undefined;
+
+    const title =
+      this._entityId === BROWSER_SOURCE
+        ? `${this.hass.localize("ui.components.media-browser.web-browser")} - `
+        : stateObj?.attributes.friendly_name
+        ? `${stateObj?.attributes.friendly_name} - `
+        : undefined;
+
+    return html`
+      <ha-app-layout>
+        <app-header fixed slot="header">
+          <app-toolbar>
+            <ha-menu-button
+              .hass=${this.hass}
+              .narrow=${this.narrow}
+            ></ha-menu-button>
+            <div main-title>
+              ${title || ""}${this.hass.localize(
+                "ui.components.media-browser.media-player-browser"
+              )}
+            </div>
+            <mwc-icon-button
+              .label=${this.hass.localize(
+                "ui.components.media-browser.choose-player"
+              )}
+              @click=${this._showMediaSouceDialog}
+            >
+              <ha-svg-icon .path=${mdiPlayNetwork}></ha-svg-icon>
+            </mwc-icon-button>
+          </app-toolbar>
+        </app-header>
+        <div class="content">
+          <ha-media-player-browse
+            .hass=${this.hass}
+            .entityId=${this._entityId}
+            @media-picked=${this._mediaPicked}
+          ></ha-media-player-browse>
+        </div>
+      </ha-app-layout>
+    `;
+  }
+
+  private _showMediaSouceDialog(): void {
+    showSelectMediaSourceDialog(this, {
+      mediaSources: this._browseMediaSources(this.hass),
+      sourceSelectedCallback: (entityId) => {
+        this._entityId = entityId;
+        this.requestUpdate();
+      },
+    });
+  }
+
+  private async _mediaPicked(
+    ev: HASSDomEvent<MediaPickedEvent>
+  ): Promise<void> {
+    const item = ev.detail.item;
+    if (this._entityId === BROWSER_SOURCE) {
+      const resolvedUrl: any = await this.hass.callWS({
+        type: "media_source/resolve_media",
+        media_content_id: item.media_content_id,
+      });
+
+      showMediaPlayerDialog(this, {
+        sourceUrl: resolvedUrl.url,
+        sourceType: resolvedUrl.mime_type,
+        title: item.title,
+      });
+      return;
+    }
+
+    this.hass!.callService("media_player", "play_media", {
+      entity_id: this._entityId,
+      media_content_id: item.media_content_id,
+      media_content_type: item.media_content_type,
+    });
+  }
+
+  private _browseMediaSources = memoize((hass: HomeAssistant) => {
```

Comment:
A memoize with just `hass` as parameter doesn't do much, as it will update everytime hass changes, and that is a LOT.

Reply:
Would be better to do this in updated and check if there is a new/removed `media_player` entity. Like for example: https://github.com/home-assistant/frontend/blob/dev/src/panels/config/zone/ha-config-zone.ts#L296-L314

Reply:
But since we only call it when we open the dialog, we don't need memoize.

Reply:
WebBrowserPlayMedia